cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractMetadataStoreMessagingSubscriberService.java

```diff
@@ -0,0 +1,115 @@
+package io.cdap.cdap.metadata;
+
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.messaging.Message;
+import io.cdap.cdap.api.messaging.MessagingContext;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.store.AppMetadataStore;
+import io.cdap.cdap.messaging.MessagingService;
+import io.cdap.cdap.messaging.context.MultiThreadMessagingContext;
+import io.cdap.cdap.messaging.subscriber.AbstractMessagingSubscriberService;
+import io.cdap.cdap.proto.codec.EntityIdTypeAdapter;
+import io.cdap.cdap.proto.id.EntityId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.TableNotFoundException;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import org.apache.tephra.TxConstants;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+/**
+ * An abstract base class for implementing metadata event message consumption from TMS.
+ *
+ * @param <T> the type that each message will be decoded to.
+ */
+public abstract class AbstractMetadataStoreMessagingSubscriberService<T> extends AbstractMessagingSubscriberService<T> {
+  private final CConfiguration cConf;
+  private final MultiThreadMessagingContext messagingContext;
+  private final TransactionRunner transactionRunner;
+  private final int maxRetriesOnConflict;
+  private final MetricsCollectionService metricsCollectionService;
+  private final Class<T> messageClass;
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+    .create();
+
+  AbstractMetadataStoreMessagingSubscriberService(CConfiguration cConf, MessagingService messagingService,
+                                                  MetricsCollectionService metricsCollectionService,
+                                                  TransactionRunner transactionRunner, Class<T> messageClass) {
+    super(
+      NamespaceId.SYSTEM.topic(cConf.get(Constants.AppFabric.APP_DELETION_EVENT_TOPIC)),
```

Comment:
@tivv @albertshau Should we rename these constants to be more generic, something like APP_METADATA_EVENT... ? Or should we change the class name?

Reply:
If it's an abstract service, there should be nothing about APP_DELETE in this class.

Reply:
Overall, I'd recommend to try using this abstract class for other children of AbstractMessagingSubscriberService, e.g. make `MetadataSubscriberService` extend from it. This would give you a proper common denominator and what should be constants and what should be parameters

Reply:
created https://cdap.atlassian.net/browse/CDAP-20419

Reply:
But it makes little sense as it is now.... I am still fine leaving it as is given the CDAP-20419 is a blocker for 6.9.0 ( updated the status)

Reply:
I will update the constants in this PR. CDAP-20419  will handle the other classes abstracting `AbstractMetadataStoreMessagingSubscriberService`